release/releaser.go

```diff
@@ -44,233 +42,376 @@ func NewReleaser(
+	include := flux.ServiceIDSet{}
+	include.Add(ss)
+	return rc.SelectServices(include, locked, excluded, logStatus)
+}
 
-	var (
-		base  = r.metrics.StageDuration.With("method", method)
-		stage *metrics.Timer
-	)
+func (rc *ReleaseContext) FindDefinedServices() ([]*ServiceUpdate, error) {
+	services, err := kubernetes.FindDefinedServices(rc.RepoPath())
+	if err != nil {
+		return nil, err
+	}
 
-	defer func() { stage.ObserveDuration() }()
-	stage = metrics.NewTimer(base.With("stage", "fetch_platform_services"))
+	var defined []*ServiceUpdate
+	for id, paths := range services {
+		switch len(paths) {
+		case 1:
+			def, err := ioutil.ReadFile(paths[0])
+			if err != nil {
+				return nil, err
+			}
+			defined = append(defined, &ServiceUpdate{
+				ServiceID:     id,
+				ManifestPath:  paths[0],
+				ManifestBytes: def,
+			})
+		default:
+			return nil, fmt.Errorf("multiple resource files found for service %s: %s", id, strings.Join(paths, ", "))
+		}
+	}
+	return defined, nil
+}
 
-	services, err := getServices.SelectServices(inst)
-	if err != nil {
-		return nil, errors.Wrap(err, "fetching platform services")
+// --- /Additions
+
+// Find all the image updates that should be performed, and do
+// replacements. (For a dry-run, we don't strictly need to do the
+// replacements, since we won't be committing any changes back;
+// however we do want to see if we *can* do the replacements, because
+// if not, it indicates there's likely some problem with the running
+// system vs the definitions given in the repo.)
+func (r *Releaser) calculateImageUpdates(rc *ReleaseContext, updates []*ServiceUpdate, releaseJob *jobs.ReleaseJobParams, logStatus statusFn) ([]*ServiceUpdate, error) {
+	var images instance.ImageMap
+	var err error
+
+	switch releaseJob.ImageSpec {
+	case flux.ImageSpecNone:
+		images = instance.ImageMap{}
+	case flux.ImageSpecLatest:
+		images, err = rc.CollectAvailableImages(updates)
+	default:
+		images, err = rc.Instance.ExactImages([]flux.ImageID{flux.ImageID(releaseJob.ImageSpec)})
 	}
-	if len(services) == 0 {
-		res = append(res, r.releaseActionPrintf("No selected services found. Nothing to do."))
-		return res, nil
+
+	if err != nil {
+		return nil, err
 	}
 
-	stage.ObserveDuration()
-	stage = metrics.NewTimer(base.With("stage", "finalize"))
+	// Do all the updates that could be written out
+	return rc.CalculateContainerUpdates(updates, images, logStatus)
+}
 
-	res = append(res, r.releaseActionPrintf(msg))
-	res = append(res, r.releaseActionClone())
+// execute is called if the release is *not* a dry run; i.e., it
+// should actually effect the changes calculated.
+func (r *Releaser) execute(rc *ReleaseContext, updates []*ServiceUpdate, releaseJob *jobs.ReleaseJobParams, logStatus statusFn) error {
+	// Write changes, commit and push
+	logStatus("Writing changes to files.")
+	err := rc.WriteUpdates(updates)
+	if err != nil {
+		return err
+	}
```

Comment:
Need to differentiate between:

a) there are no updates. (bail, and noop)
b) we're applying latest config (no commit & push to do, but continue)